File: Dart/merge_intervals.dart
import 'dart:math';

class IntervalPart implements Comparable<IntervalPart> {
  final int index;
  final int value;
  final bool isStart;

  IntervalPart(this.index, this.value, this.isStart);

  @override
  int compareTo(IntervalPart other) {
    if (value != other.value) {
      return value.compareTo(other.value);
    } else {
      return isStart == other.isStart ? 0 : (isStart ? -1 : 1);
    }
  }
}


class Solution {
  List<List<int>> merge(List<List<int>> intervals) {
    int length = intervals.length;
    if (length == 1) return intervals;
    List<List<int>> result = [];
    List<IntervalPart> intervalsPartList = List.generate(
        length * 2,
        (int index) => IntervalPart(
            index ~/ 2, intervals[index ~/ 2][index % 2], index % 2 == 0));
    intervalsPartList.sort();

    int stack = 1;
    List<int> interval = [intervalsPartList[0].value];

    for (int i = 1; i < length * 2; i++) {
      if (intervalsPartList[i].isStart) {
        stack++;
      } else {
        stack=max(stack-1,0);
      }
      if (stack == 0) {
        interval.add(intervalsPartList[i].value);
        result.add(interval);
        if (i != 2 * length - 1) {
          interval = [intervalsPartList[i + 1].value];
          stack = 1;
          i++;
        }
      }
    }
    return result;
  }
}

File: Python/merge-intervals.py
from typing import List


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:

        intervals.sort(key=lambda x: x[0])
        result: List[List[int]] = [intervals[0]]

        for l in intervals[1:]:
            current = result[-1]
            if current[0] <= l[0] <= current[1]:
                if current[1] < l[1]:
                    current[1] = l[1]
            else:
                result.append(l)
        return result


if __name__ == "__main__":
    intervals = [[1, 3], [2, 6], [8, 10], [15, 18]]
    sol = Solution()
    result = sol.merge(intervals)
    print(result)
